relays/ethereum/src/ethereum_exchange.rs

```diff
@@ -0,0 +1,212 @@
+			None => return Ok(None),
+		};
+
+		// we need transaction to be mined => check if it is included in the block
+		let eth_header_id = match (eth_tx.block_number, eth_tx.block_hash) {
+			(Some(block_number), Some(block_hash)) => HeaderId(block_number.as_u64(), block_hash),
+			_ => return Ok(None),
+		};
+
+		Ok(Some((eth_header_id, eth_tx)))
+	}
+
+	async fn transaction_proof(
+		&self,
+		eth_header_id: &EthereumHeaderId,
+		eth_tx: EthereumTransaction,
+	) -> Result<EthereumTransactionInclusionProof, Self::Error> {
+		const TRANSACTION_HAS_RAW_FIELD_PROOF: &'static str = "RPC level checks that transactions from Ethereum\
+			node are having `raw` field; qed";
+
+		let eth_header = self.client.header_by_hash_with_transactions(eth_header_id.1).await?;
+		let eth_relay_tx_hash = eth_tx.hash;
+		let mut eth_relay_tx = Some(eth_tx);
+		let mut eth_relay_tx_index = None;
+		let mut transaction_proof = Vec::with_capacity(eth_header.transactions.len());
+		for (index, eth_tx) in eth_header.transactions.into_iter().enumerate() {
+			if eth_tx.hash != eth_relay_tx_hash {
+				let eth_raw_tx = eth_tx.raw.expect(TRANSACTION_HAS_RAW_FIELD_PROOF);
+				transaction_proof.push(eth_raw_tx.0);
+			} else {
+				let eth_raw_relay_tx = match eth_relay_tx.take() {
+					Some(eth_relay_tx) => eth_relay_tx.raw.expect(TRANSACTION_HAS_RAW_FIELD_PROOF),
+					None => {
+						return Err(
+							EthereumNodeError::DuplicateBlockTransaction(*eth_header_id, eth_relay_tx_hash).into(),
+						)
+					}
+				};
+				eth_relay_tx_index = Some(index as u64);
+				transaction_proof.push(eth_raw_relay_tx.0);
+			}
+		}
+
+		Ok(EthereumTransactionInclusionProof {
+			block: eth_header_id.1,
+			index: eth_relay_tx_index.ok_or_else(|| {
+				RpcError::from(EthereumNodeError::BlockMissingTransaction(
+					*eth_header_id,
+					eth_relay_tx_hash,
+				))
+			})?,
+			proof: transaction_proof,
+		})
+	}
+}
+
+/// Substrate node as transactions proof target.
+struct SubstrateTransactionsTarget {
+	client: SubstrateRpcClient,
+	sign_params: SubstrateSigningParams,
+}
+
+#[async_trait]
+impl TargetClient<EthereumToSubstrateExchange> for SubstrateTransactionsTarget {
+	type Error = RpcError;
+
+	async fn tick(&self) {
+		async_std::task::sleep(SUBSTRATE_TICK_INTERVAL).await;
+	}
+
+	async fn is_header_known(&self, id: &EthereumHeaderId) -> Result<bool, Self::Error> {
+		self.client.ethereum_header_known(*id).await
+	}
+
+	async fn is_header_finalized(&self, id: &EthereumHeaderId) -> Result<bool, Self::Error> {
+		// we check if header is finalized by simple comparison of the header' number and
+		// number of best finalized PoA' header known to Substrate node.
+		//
+		// this may lead to failure in tx proof import if PoA reorganization has happened
+		// after we have checked that our tx has been included into given block
+		//
+		// the fix is easy, but since this code is mostly developed for demonstration purposes,
+		// I'm leaving this KISS-based design here
```

Comment:
In 10 years some poor soul gonna be shaking their fist at this "demonstration" code ;)